File: companion/ai_translator.py
"""
AI Shell Command Translator for CmdBar Companion & CLI.
Translates natural language prompts to executable shell commands.
Supports OpenAI, Anthropic (Claude), and Ollama (local model), with fallback.
"""

import os
import re
import json
import urllib.request
import urllib.error


def parse_command_from_ai_response(response_text: str) -> str:
    """
    Extracts executable command block from LLM response text.
    """
    if not response_text or not isinstance(response_text, str):
        return ""
    
    text = response_text.strip()

    # 1. Markdown code block ```bash ... ``` or ``` ... ```
    match_block = re.search(r"```(?:[a-zA-Z0-9_-]+)?\s*\n?([\s\S]*?)\n?```", text)
    if match_block and match_block.group(1).strip():
        return match_block.group(1).strip()

    # 2. Inline backticks `...`
    match_inline = re.search(r"`([^`\n]+)`", text)
    if match_inline and match_inline.group(1).strip():
        return match_inline.group(1).strip()

    # 3. Fallback: line-by-line filtering
    lines = [l.strip() for l in text.split("\n") if l.strip()]
    clean_lines = []
    for line in lines:
        lower = line.lower()
        if not (lower.startswith("here is") or lower.startswith("you can") or 
                lower.startswith("note:") or lower.startswith("explanation:") or
                lower.startswith("command:")):
            clean_lines.append(line)

    return " ".join(clean_lines).strip() or text


def clean_ai_prompt(prompt: str) -> str:
    """
    Strips '/ai ' prefix from user prompt.
    """
    if not prompt or not isinstance(prompt, str):
        return ""
    return re.sub(r"^\s*\/ai\s*", "", prompt, flags=re.IGNORECASE).strip()


def is_ai_command(text: str) -> bool:
    """
    Checks if command string starts with '/ai' trigger.
    """
    if not text or not isinstance(text, str):
        return False
    return text.strip().lower().startswith("/ai")


def get_ai_api_key(provider: str, config: dict = None) -> str:
    """
    Retrieves API key for provider from config or environment variables.
    """
    config = config or {}
    ai_cfg = config.get("ai", config) if isinstance(config, dict) else {}

    key = ai_cfg.get("api_key") or ai_cfg.get("apiKey")
    if key and isinstance(key, str) and key.strip():
        return key.strip()

    prov = (provider or ai_cfg.get("provider") or "openai").lower()
    if prov == "openai":
        return os.environ.get("OPENAI_API_KEY", "").strip()
    elif prov in ("anthropic", "claude"):
        return os.environ.get("ANTHROPIC_API_KEY", "").strip()
    elif prov == "ollama":
        return os.environ.get("OLLAMA_API_KEY", "").strip()

    return ""


def build_ai_request(provider: str, raw_prompt: str, options: dict = None) -> tuple:
    """
    Builds HTTP request tuple (url, headers, body_bytes, provider) for provider.
    """
    options = options or {}
    prov = (provider or "openai").lower()
    prompt = clean_ai_prompt(raw_prompt)
    system_prompt = options.get(
        "system_prompt",
        "You are an expert Linux shell command assistant. Translate the natural language prompt into a concise, correct executable shell command. Output ONLY the raw executable command inside a ```bash ... ``` code block. Do not include explanation or Markdown outside the code block."
    )
    temp = options.get("temperature", 0.2)
    api_key = get_ai_api_key(prov, options)

    if prov == "openai":
        endpoint = options.get("endpoint") or "https://api.openai.com/v1/chat/completions"
        model = options.get("model") or "gpt-4o"
        headers = {"Content-Type": "application/json"}
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"
        body = {
            "model": model,
            "temperature": temp,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": prompt}
            ]
        }
        return endpoint, headers, json.dumps(body).encode("utf-8"), "openai"

    elif prov in ("anthropic", "claude"):
        endpoint = options.get("endpoint") or "https://api.anthropic.com/v1/messages"
        model = options.get("model") or "claude-3-5-sonnet-20241022"
        headers = {
            "Content-Type": "application/json",
            "anthropic-version": "2023-06-01"
        }
        if api_key:
            headers["x-api-key"] = api_key
        body = {
            "model": model,
            "max_tokens": 1024,
            "temperature": temp,
            "messages": [
                {"role": "user", "content": f"{system_prompt}\n\nUser Prompt: {prompt}"}
            ]
        }
        return endpoint, headers, json.dumps(body).encode("utf-8"), "anthropic"

    elif prov == "ollama":
        endpoint = options.get("endpoint") or "http://localhost:11434/api/generate"
        model = options.get("model") or "llama3"
        headers = {"Content-Type": "application/json"}
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"

        if "/chat/completions" in endpoint:
            body = {
                "model": model,
                "temperature": temp,
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": prompt}
                ]
            }
        else:
            body = {
                "model": model,
                "prompt": f"{system_prompt}\n\nUser Prompt: {prompt}",
                "stream": False,
                "options": {"temperature": temp}
            }
        return endpoint, headers, json.dumps(body).encode("utf-8"), "ollama"

    else:
        raise ValueError(f"Unsupported AI provider: {provider}")


def extract_ai_response_text(provider: str, response_data: dict) -> str:
    """
    Parses provider response dict to extract text output.
    """
    if not isinstance(response_data, dict):
        return ""
    prov = (provider or "").lower()

    if prov == "openai":
        choices = response_data.get("choices", [])
        if choices and isinstance(choices, list):
            return choices[0].get("message", {}).get("content", "")
    elif prov in ("anthropic", "claude"):
        content = response_data.get("content")
        if isinstance(content, list):
            return "\n".join(c.get("text", "") for c in content if isinstance(c, dict))
        elif isinstance(content, str):
            return content
    elif prov == "ollama":
        if "response" in response_data:
            return response_data["response"]
        choices = response_data.get("choices", [])
        if choices and isinstance(choices, list):
            return choices[0].get("message", {}).get("content", "")

    return (
        response_data.get("response") or
        response_data.get("text") or
        ""
    )


def http_post_json(url: str, headers: dict, data_bytes: bytes, timeout: int = 15) -> dict:
    """
    Executes HTTP POST using urllib.request.
    """
    req = urllib.request.Request(url, data=data_bytes, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            resp_body = resp.read().decode("utf-8")
            return json.loads(resp_body)
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8") if e.fp else ""
        raise RuntimeError(f"HTTP {e.code}: {e.reason} - {err_body}")
    except Exception as e:
        raise RuntimeError(f"HTTP Request failed: {str(e)}")


def translate_natural_language_to_command(raw_prompt: str, config: dict = None) -> str:
    """
    Translates natural language prompt to executable shell command.
    Includes automatic local model fallback if primary provider fails.
    """
    prompt = clean_ai_prompt(raw_prompt)
    if not prompt:
        raise ValueError("Prompt cannot be empty.")

    config = config or {}
    ai_cfg = config.get("ai", config) if isinstance(config, dict) else {}
    primary_provider = (ai_cfg.get("provider") or "openai").lower()
    fallback_provider = (ai_cfg.get("fallback_provider") or "ollama").lower()

    # Try Primary Provider
    try:
        url, headers, body, prov = build_ai_request(primary_provider, prompt, ai_cfg)
        res = http_post_json(url, headers, body)
        text = extract_ai_response_text(prov, res)
        cmd = parse_command_from_ai_response(text)
        if cmd:
            return cmd
        raise RuntimeError("AI provider returned empty command.")
    except Exception as primary_err:
        if primary_provider == fallback_provider:
            raise primary_err

        # Try Fallback Provider (e.g. Ollama)
        try:
            fallback_opts = dict(ai_cfg)
            fallback_opts["model"] = ai_cfg.get("fallback_model") or ("llama3" if fallback_provider == "ollama" else ai_cfg.get("model"))
            url, headers, body, prov = build_ai_request(fallback_provider, prompt, fallback_opts)
            res = http_post_json(url, headers, body)
            text = extract_ai_response_text(prov, res)
            cmd = parse_command_from_ai_response(text)
            if cmd:
                return cmd
            raise RuntimeError("Fallback AI provider returned empty command.")
        except Exception as fallback_err:
            raise RuntimeError(
                f"AI Command Translation failed. Primary ({primary_provider}): {str(primary_err)}; Fallback ({fallback_provider}): {str(fallback_err)}"
            )

File: extension/aiTranslator.js
/**
 * AI Command Translator module for CmdBar extension.
 * Translates natural language prompts to executable shell commands.
 * Supports OpenAI, Anthropic (Claude), and Ollama (local model), with fallback.
 */

let GLib, Gio;
const isNode = typeof process !== "undefined" && process.versions && process.versions.node;

if (!isNode) {
  try {
    const gi = await import("gi");
    GLib = gi.GLib;
    Gio = gi.Gio;
  } catch (e) {}
}

/**
 * Extracts raw shell command from LLM response text.
 * @param {string} responseText
 * @returns {string}
 */
export function parseCommandFromAIResponse(responseText) {
  if (!responseText || typeof responseText !== "string") {
    return "";
  }
  const text = responseText.trim();

  // 1. Markdown code block ```bash ... ``` or ```sh ... ``` or ``` ... ```
  const codeBlockRegex = /```(?:[a-zA-Z0-9_-]+)?\s*\n?([\s\S]*?)\n?```/;
  const matchBlock = codeBlockRegex.exec(text);
  if (matchBlock && matchBlock[1] && matchBlock[1].trim()) {
    return matchBlock[1].trim();
  }

  // 2. Inline backticks `...`
  const inlineRegex = /`([^`\n]+)`/;
  const matchInline = inlineRegex.exec(text);
  if (matchInline && matchInline[1] && matchInline[1].trim()) {
    return matchInline[1].trim();
  }

  // 3. Fallback: filter line by line
  const lines = text
    .split("\n")
    .map((l) => l.trim())
    .filter(Boolean);

  const cleanLines = lines.filter((line) => {
    const lower = line.toLowerCase();
    return (
      !lower.startsWith("here is") &&
      !lower.startsWith("you can") &&
      !lower.startsWith("note:") &&
      !lower.startsWith("explanation:") &&
      !lower.startsWith("command:")
    );
  });

  return cleanLines.join(" ").trim() || text;
}

/**
 * Strips trigger prefix (e.g. "/ai ") from prompt string.
 * @param {string} prompt
 * @returns {string}
 */
export function cleanAIPrompt(prompt) {
  if (!prompt || typeof prompt !== "string") {
    return "";
  }
  return prompt.trim().replace(/^\/ai\s*/i, "").trim();
}

/**
 * Checks if input starts with AI mode prefix.
 * @param {string} text
 * @returns {boolean}
 */
export function isAICommand(text) {
  if (!text || typeof text !== "string") {
    return false;
  }
  return text.trim().toLowerCase().startsWith("/ai");
}

/**
 * Gets API key for specified provider from config or environment variables.
 * @param {string} provider
 * @param {object} [config]
 * @returns {string}
 */
export function getAIApiKey(provider, config = {}) {
  const aiCfg = config.ai || config || {};
  if (aiCfg.api_key && typeof aiCfg.api_key === "string" && aiCfg.api_key.trim()) {
    return aiCfg.api_key.trim();
  }
  if (aiCfg.apiKey && typeof aiCfg.apiKey === "string" && aiCfg.apiKey.trim()) {
    return aiCfg.apiKey.trim();
  }

  const prov = (provider || aiCfg.provider || "openai").toLowerCase();

  let envVal = "";
  if (isNode) {
    if (prov === "openai") envVal = process.env.OPENAI_API_KEY || "";
    else if (prov === "anthropic" || prov === "claude") envVal = process.env.ANTHROPIC_API_KEY || "";
    else if (prov === "ollama") envVal = process.env.OLLAMA_API_KEY || "";
  } else {
    try {
      if (typeof GLib !== "undefined" && GLib.getenv) {
        if (prov === "openai") envVal = GLib.getenv("OPENAI_API_KEY") || "";
        else if (prov === "anthropic" || prov === "claude") envVal = GLib.getenv("ANTHROPIC_API_KEY") || "";
        else if (prov === "ollama") envVal = GLib.getenv("OLLAMA_API_KEY") || "";
      }
    } catch (e) {}
  }

  return envVal ? envVal.trim() : "";
}

/**
 * Universal HTTP POST helper supporting Node.js fetch and GJS / curl fallback.
 * @param {string} url
 * @param {object} headers
 * @param {object} body
 * @returns {Promise<object>}
 */
export async function httpPost(url, headers = {}, body = {}) {
  const jsonBody = JSON.stringify(body);

  // 1. Try global fetch if available (Node.js or modern GJS)
  if (typeof fetch === "function") {
    const res = await fetch(url, {
      method: "POST",
      headers,
      body: jsonBody,
    });
    if (!res.ok) {
      const errText = await res.text().catch(() => "");
      throw new Error(`HTTP ${res.status}: ${res.statusText} - ${errText}`);
    }
    return await res.json();
  }

  // 2. Node.js fallback using child_process curl if fetch unavailable
  if (isNode) {
    const { execFile } = await import("child_process");
    const util = await import("util");
    const execFilePromise = util.promisify(execFile);

    const args = ["-s", "-S", "-X", "POST", url];
    for (const [k, v] of Object.entries(headers)) {
      args.push("-H", `${k}: ${v}`);
    }
    args.push("-d", jsonBody);

    const { stdout, stderr } = await execFilePromise("curl", args);
    if (!stdout && stderr) {
      throw new Error(`Curl request failed: ${stderr}`);
    }
    return JSON.parse(stdout);
  }

  // 3. GJS environment fallback using Gio.Subprocess curl
  if (typeof Gio !== "undefined" && Gio.Subprocess) {
    const args = ["curl", "-s", "-S", "-X", "POST", url];
    for (const [k, v] of Object.entries(headers)) {
      args.push("-H", `${k}: ${v}`);
    }
    args.push("-d", jsonBody);

    return new Promise((resolve, reject) => {
      try {
        const proc = Gio.Subprocess.new(
          args,
          Gio.SubprocessFlags.STDOUT_PIPE | Gio.SubprocessFlags.STDERR_PIPE
        );
        proc.communicate_utf8_async(null, null, (subprocess, result) => {
          try {
            const [stdout, stderr] = subprocess.communicate_utf8_finish(result);
            if (!subprocess.get_successful()) {
              reject(new Error(`Curl error: ${stderr || "Process failed"}`));
              return;
            }
            resolve(JSON.parse(stdout));
          } catch (e) {
            reject(e);
          }
        });
      } catch (e) {
        reject(e);
      }
    });
  }

  throw new Error("No available HTTP client environment found.");
}

/**
 * Builds API request configuration object for provider.
 * @param {string} provider
 * @param {string} rawPrompt
 * @param {object} [options]
 * @returns {object}
 */
export function buildAIRequest(provider, rawPrompt, options = {}) {
  const prov = (provider || "openai").toLowerCase();
  const prompt = cleanAIPrompt(rawPrompt);
  const systemPrompt =
    options.systemPrompt ||
    "You are an expert Linux shell command assistant. Translate the natural language prompt into a concise, correct executable shell command. Output ONLY the raw executable command inside a ```bash ... ``` code block. Do not include explanation or Markdown outside the code block.";
  const temp = options.temperature ?? 0.2;
  const apiKey = getAIApiKey(prov, options);

  if (prov === "openai") {
    const endpoint = options.endpoint || "https://api.openai.com/v1/chat/completions";
    const model = options.model || "gpt-4o";
    const headers = {
      "Content-Type": "application/json",
    };
    if (apiKey) {
      headers["Authorization"] = `Bearer ${apiKey}`;
    }
    const body = {
      model,
      temperature: temp,
      messages: [
        { role: "system", content: systemPrompt },
        { role: "user", content: prompt },
      ],
    };
    return { endpoint, headers, body, provider: "openai" };
  } else if (prov === "anthropic" || prov === "claude") {
    const endpoint = options.endpoint || "https://api.anthropic.com/v1/messages";
    const model = options.model || "claude-3-5-sonnet-20241022";
    const headers = {
      "Content-Type": "application/json",
      "anthropic-version": "2023-06-01",
    };
    if (apiKey) {
      headers["x-api-key"] = apiKey;
    }
    const body = {
      model,
      max_tokens: 1024,
      temperature: temp,
      messages: [
        { role: "user", content: `${systemPrompt}\n\nUser Prompt: ${prompt}` },
      ],
    };
    return { endpoint, headers, body, provider: "anthropic" };
  } else if (prov === "ollama") {
    let endpoint = options.endpoint || "http://localhost:11434/api/generate";
    const model = options.model || "llama3";
    const headers = {
      "Content-Type": "application/json",
    };
    if (apiKey) {
      headers["Authorization"] = `Bearer ${apiKey}`;
    }

    if (endpoint.includes("/chat/completions")) {
      const body = {
        model,
        temperature: temp,
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user", content: prompt },
        ],
      };
      return { endpoint, headers, body, provider: "ollama" };
    } else {
      const body = {
        model,
        prompt: `${systemPrompt}\n\nUser Prompt: ${prompt}`,
        stream: false,
        options: { temperature: temp },
      };
      return { endpoint, headers, body, provider: "ollama" };
    }
  } else {
    throw new Error(`Unsupported AI provider: ${provider}`);
  }
}

/**
 * Extracts response text from provider JSON output.
 * @param {string} provider
 * @param {object} responseData
 * @returns {string}
 */
export function extractAIResponseText(provider, responseData) {
  if (!responseData || typeof responseData !== "object") {
    return "";
  }
  const prov = (provider || "").toLowerCase();

  if (prov === "openai") {
    return responseData.choices?.[0]?.message?.content || "";
  } else if (prov === "anthropic" || prov === "claude") {
    if (Array.isArray(responseData.content)) {
      return responseData.content.map((c) => c.text || "").join("\n");
    }
    return responseData.content || "";
  } else if (prov === "ollama") {
    if (responseData.response) {
      return responseData.response;
    }
    if (responseData.choices?.[0]?.message?.content) {
      return responseData.choices[0].message.content;
    }
  }

  return (
    responseData.response ||
    responseData.text ||
    responseData.choices?.[0]?.message?.content ||
    ""
  );
}

/**
 * Main translation function with local model fallback support.
 * @param {string} rawPrompt
 * @param {object} [config]
 * @returns {Promise<string>}
 */
export async function translateNaturalLanguageToCommand(rawPrompt, config = {}) {
  const prompt = cleanAIPrompt(rawPrompt);
  if (!prompt) {
    throw new Error("Prompt cannot be empty.");
  }

  const aiConfig = config.ai || config || {};
  const primaryProvider = (aiConfig.provider || "openai").toLowerCase();
  const fallbackProvider = (aiConfig.fallback_provider || "ollama").toLowerCase();

  // Try Primary Provider
  try {
    const req = buildAIRequest(primaryProvider, prompt, {
      ...aiConfig,
      apiKey: getAIApiKey(primaryProvider, config),
    });
    const res = await httpPost(req.endpoint, req.headers, req.body);
    const text = extractAIResponseText(primaryProvider, res);
    const command = parseCommandFromAIResponse(text);
    if (command) {
      return command;
    }
    throw new Error("AI provider returned empty command.");
  } catch (primaryErr) {
    console.warn(
      `CmdBar AI: Primary provider (${primaryProvider}) failed: ${primaryErr.message}. Attempting fallback (${fallbackProvider})...`
    );

    if (primaryProvider === fallbackProvider) {
      throw primaryErr;
    }

    // Try Fallback Provider (e.g. Ollama)
    try {
      const fallbackOptions = {
        ...aiConfig,
        model: aiConfig.fallback_model || (fallbackProvider === "ollama" ? "llama3" : aiConfig.model),
        apiKey: getAIApiKey(fallbackProvider, config),
      };
      const req = buildAIRequest(fallbackProvider, prompt, fallbackOptions);
      const res = await httpPost(req.endpoint, req.headers, req.body);
      const text = extractAIResponseText(fallbackProvider, res);
      const command = parseCommandFromAIResponse(text);
      if (command) {
        return command;
      }
      throw new Error("Fallback AI provider returned empty command.");
    } catch (fallbackErr) {
      throw new Error(
        `AI Command Translation failed. Primary (${primaryProvider}): ${primaryErr.message}; Fallback (${fallbackProvider}): ${fallbackErr.message}`
      );
    }
  }
}
